### PR TITLE
Common alerts

### PR DESCRIPTION
**Pull request by Eabramov84**

### CODE REVIEW

File: alerts/alerts.go [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+
+	instance.journal.Close()
+}
+
+// GetAlertsChannel returns channel with alerts to be sent.
+func (instance *Alerts) GetAlertsChannel() (channel <-chan cloudprotocol.AlertItem) {
+	return instance.alertsChannel
+}
+
+// SendResourceAlert sends resource alert.
```
Comment (al1img) [inline alerts/alerts.go:214]: Fix comment

**Comment by al1img** [comment]
In general, it doesn't look good from architecture point of view. A better solution could be:
1. Move journal handling functionality into aos_common journalalerts module.
2. SM:
    a. create journalalerts instance if configured;
    b. all modules should send alerts directly to smserver.
3. CM:
    a. create journalalerts instance if configured;
    b.  all modules should send alerts directly to alerts package that aggregates them and send to amqp.

File: alerts/alerts.go [hunk 2]
```diff
@@ -47,6 +48,7 @@ const aosServicePrefix = "aos-service@"
 const (
 	waitJournalTimeout = 1 * time.Second
 	journalSavePeriod  = 10 * time.Second
+	allocAlertItems    = 10
```
[the remark this answers is not in the document]
Comment (Eabramov84) [reply alerts/alerts.go]: removed

File: journalalerts/journalalerts.go [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+)
+
+const microSecondsInSecond = 1000000
+
+/***********************************************************************************************************************
+ * Types
+ **********************************************************************************************************************/
+
+// Sender alerts sender to the cloud interface.
+type Sender interface {
```
Comment (al1img) [inline journalalerts/journalalerts.go:56]: `AlertSender`?

File: journalalerts/journalalerts.go [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+		}
+
+		alert := cloudprotocol.AlertItem{
+			Timestamp: time.Unix(int64(entry.RealtimeTimestamp/microSecondsInSecond),
+				int64((entry.RealtimeTimestamp%microSecondsInSecond)*1000)),
+		}
+
+		instance.fillServiceInstanceAlert(&alert, entry, unit)
+
+		if alert.Payload == nil {
```
Comment (al1img) [inline journalalerts/journalalerts.go:312]: It is not clear that fill.. function change alert.Payload inside. You should look inside function to understand what it does. I suggest the following:
```go
if alert.Payload = getServiceInstanceAlert(...); alert.payload != nil {
    alert.Tag = cloudprotocol.AlertTagServiceInstance
} else if alert.Payload = get...; alert.Payload != nil {
    alert.Tag = ...
} else {
    continue
}

instance.sender.SendAlert(alert)
```

get... function return nil or found alert type instance.

File: journalalerts/journalalerts.go [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+}
+
+func (instance *JournalAlerts) fillSystemAlert(alert *cloudprotocol.AlertItem, entry *sdjournal.JournalEntry) {
+	skipped := false
+
+	for _, substr := range instance.filterRegexp {
+		skipped = substr.MatchString(entry.Fields[sdjournal.SD_JOURNAL_FIELD_MESSAGE])
+
+		if skipped {
+			break
```
Comment (al1img) [inline journalalerts/journalalerts.go:394]: if substr.MatchString {
    return
}

**Review by al1img — APPROVED** [review]

File: journalalerts/journalalerts.go [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+
+	for {
+		select {
+		case <-journalTicker.C:
+			if err := instance.storeCurrentCursor(); err != nil {
+				log.Error("Can't store journal cursor: ", err)
+			}
+
+		case <-ctx.Done():
+			journalTicker.Stop()
```
Comment (MykolaSolyanko) [inline journalalerts/journalalerts.go:243]: maybe in this case we should `return` ? to avoid possible goroutine leak

**Review by MykolaSolyanko — APPROVED** [review]

File: journalalerts/journalalerts.go (hunk 6 above)
Comment (Eabramov84) [reply journalalerts/journalalerts.go:243]: fixed

**Pull request merged**